### PR TITLE
Prove liveness for rabbitmq

### PR DESCRIPTION
**Pull request by euclidgame**
Because the rabbitmq controller doesn't support scaling down, the current liveness property we target is `spec |= []cr_exists ~> [] config_map_exists`.

### CODE REVIEW

File: src/controller_examples/zookeeper_controller/proof/safety.rs [hunk 1]
```diff
@@ -116,18 +116,13 @@ pub open spec fn filtered_create_sts_req_len_is_at_most_one(
...
-            if s.pending_req_of(key).content.get_rest_id() >= rest_id {
-                ||| {
-                    &&& s.message_in_flight(s.pending_req_of(key))
-                    &&& s.network_state.in_flight.filter(sts_create_request_msg_since(key, rest_id)).len() == 1
-                }
-                ||| s.network_state.in_flight.filter(sts_create_request_msg_since(key, rest_id)).len() == 0
-            } else {
-                s.network_state.in_flight.filter(sts_create_request_msg_since(key, rest_id)).len() == 0
-            }
+        s.network_state.in_flight.filter(sts_create_request_msg_since(key, rest_id)).len() > 0
```
Comment (marshtompsxd) [inline src/controller_examples/zookeeper_controller/proof/safety.rs:119]: This refactoring is nice!

File: src/controller_examples/zookeeper_controller/proof/safety.rs [hunk 2]
```diff
@@ -300,42 +295,15 @@ pub open spec fn at_most_one_create_sts_req_since_rest_id_is_in_flight(
...
-                    &&& #[trigger] s.network_state.in_flight.contains(msg)
-                    &&& sts_create_request_msg_since(key, rest_id)(msg)
-                }
-            } else {
-                forall |msg| !{
-                    &&& #[trigger] s.network_state.in_flight.contains(msg)
-                    &&& sts_create_request_msg_since(key, rest_id)(msg)
-                }
-            }
+        forall |msg| {
```
Comment (marshtompsxd) [inline src/controller_examples/zookeeper_controller/proof/safety.rs:298]: This refactoring is nice, too!

File: src/controller_examples/rabbitmq_controller/proof/terminate.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        spec.entails(always(lift_state(pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), rabbitmq_reconcile_state(RabbitmqReconcileStep::AfterCreateRoleBinding))))),
+        spec.entails(always(lift_state(pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), rabbitmq_reconcile_state(RabbitmqReconcileStep::AfterGetStatefulSet))))),
+        spec.entails(always(lift_state(pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), rabbitmq_reconcile_state(RabbitmqReconcileStep::AfterCreateStatefulSet))))),
+        spec.entails(always(lift_state(pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), rabbitmq_reconcile_state(RabbitmqReconcileStep::AfterUpdateStatefulSet))))),
+    ensures
+        spec.entails(
+            true_pred().leads_to(lift_state(|s: ClusterState| !s.reconcile_state_contains(rabbitmq.object_ref())))
+        ),
+{
+    let reconcile_idle = |s: ClusterState| { !s.reconcile_state_contains(rabbitmq.object_ref()) };
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/proof/terminate.rs:70]: Maybe it is better to use assert_by to explicitly show what the following lemmas are used for?

Comment (euclidgame) [reply src/controller_examples/rabbitmq_controller/proof/terminate.rs:70]: I don't like that, at least for termination. These lemmas are used very straightforwardly.

**Pull request merged**